pkg/controller/direct/asset/feed_controller.go

```diff
@@ -0,0 +1,327 @@
+	}
+	return nil
+}
+
+func (a *feedAdapter) Create(ctx context.Context, createOp *directbase.CreateOperation) error {
+	log := klog.FromContext(ctx)
+	log.Info("---------creating asset feed", "name", a.id)
+	mapCtx := &direct.MapContext{}
+
+	if err := a.normalizeReferences(ctx); err != nil {
+		return fmt.Errorf("normalizing references: %w", err)
+	}
+
+	desired := a.desired.DeepCopy()
+	resource := AssetFeedSpec_ToProto(mapCtx, &desired.Spec)
+	if mapCtx.Err() != nil {
+		return mapCtx.Err()
+	}
+
+	resource.Name = a.id.String() // Name is required for CreateFeed
+
+	req := &pb.CreateFeedRequest{
+		Parent: a.id.Parent().String(),
+		FeedId: a.id.ID(),
+		Feed:   resource,
+	}
+
+	// Attempt to create the feed
+	actual, err := a.gcpClient.CreateFeed(ctx, req)
+
+	// Regardless of CreateFeed result (success or AlreadyExists),
+	// we need to ensure a.actual is populated by calling Find.
+	found := false
+	var findErr error
+
+	if err == nil {
+		log.Info("---------successfully created asset feed in gcp, attempting to fetch it", "name", a.id)
+		a.actual = actual
+		found = true
+	} else if errors.IsAlreadyExists(err) {
+		// Resource already exists. Log and attempt to Find to populate a.actual.
+		log.Info("---------asset feed already exists during creation attempt, attempting to fetch it", "name", a.id.String(), "warning", err)
+		found, findErr = a.Find(ctx)
+		if findErr != nil {
+			return fmt.Errorf("fetching existing asset feed %q after CreateFeed returned AlreadyExists failed: %w", a.id.String(), findErr)
+		}
+		if !found {
+			// This is unexpected if CreateFeed returned AlreadyExists
+			return fmt.Errorf("asset feed %q not found even though CreateFeed returned AlreadyExists", a.id.String())
+		}
+	} else {
+		// Other error during creation
+		return fmt.Errorf("creating asset feed %s: %w", a.id.String(), err)
+	}
+
+	// Update status
+	status := &krm.AssetFeedStatus{}
+	status.ExternalRef = direct.LazyPtr(a.actual.Name)
+	// Pass the fetched 'a.actual' to UpdateStatus for potential status mapping
+	return createOp.UpdateStatus(ctx, status, nil)
+}
+
+func (a *feedAdapter) Update(ctx context.Context, updateOp *directbase.UpdateOperation) error {
+	log := klog.FromContext(ctx)
+	log.Info("---------updating asset feed", "name", a.id, "actual", a.actual.Name)
+	mapCtx := &direct.MapContext{}
+
+	if err := a.normalizeReferences(ctx); err != nil {
+		return fmt.Errorf("normalizing references: %w", err)
+	}
+
+	desired := a.desired.DeepCopy()
+	resource := AssetFeedSpec_ToProto(mapCtx, &desired.Spec)
+	if mapCtx.Err() != nil {
+		return mapCtx.Err()
+	}
+	resource.Name = a.id.String() // Name must be set for update
+
+	paths, err := common.CompareProtoMessage(resource, a.actual, common.BasicDiff)
+	if err != nil {
+		return fmt.Errorf("comparing proto messages: %w", err)
+	}
+	// Name is output only, but is required in the update request object.
+	paths.Delete("name")
```

Comment:
Yeah, `CompareProtoMessage` doesn't work if the proto fields are tagged incorrectly. But I'd expect "name" not to appear in the paths since you've already set the `name` field in `resource` (therefore both `resource` and `a.actual` should have the same name field?).

Reply:
i guess the mock and realgcp implementation were differing a bit. projectID vs projectNumber. ill comment it out and test.